plugin/validator_test.go

```diff
@@ -956,6 +956,889 @@ func TestValidator_SchemaAsJSON(t *testing.T) {
+				"semantic_version": "1.2.3-alpha.1",
+				"tag": "test"
+			}`,
+		},
+		{
+			name: "valid semantic_version",
+			config: `{
+				"semantic_version": "1.2.3-beta.1",
+				"tag": "test"
+			}`,
+		},
+		{
+			name: "valid semantic_version",
+			config: `{
+				"semantic_version": "1.2.3.4-enterprise-edition",
+				"tag": "test"
+			}`,
+		},
+		{
+			name: "invalid semantic_version",
+			config: `{
+				"semantic_version": "hello",
+				"tag": "test"
+			}`,
+			wantErr:     true,
+			expectedErr: `{"config":{"semantic_version":"invalid version number: must be in format of X.Y.Z"}}`,
+		},
+		{
+			name: "invalid semantic_version",
+			config: `{
+				"semantic_version": ".1",
+				"tag": "test"
+			}`,
+			wantErr:     true,
+			expectedErr: `{"config":{"semantic_version":"invalid version number: must be in format of X.Y.Z"}}`,
+		},
+		{
+			name: "invalid semantic_version",
+			config: `{
+				"semantic_version": "1..1",
+				"tag": "test"
+			}`,
+			wantErr:     true,
+			expectedErr: `{"config":{"semantic_version":"must not have empty version segments"}}`,
+		},
+		{
+			name: "lua code not accepted",
+			config: `{
+				"lua_code": {"header": "return nil"},
+				"tag": "test"
+			}`,
+			wantErr: true,
+			expectedErr: `{"config":{
+				"lua_code":"Error parsing function: lua-tree/share/lua/5.1/kong/tools/sandbox.lua:119: ` +
+				`loading of untrusted Lua code disabled because 'untrusted_lua' config option is set to 'off'"
+				}}`,
+		},
+	}
+
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			plugin := `{
+				"name": "all-typedefs",
+				"enabled": true,
+				"protocols": ["http"],
+				"config": %s
+			}`
+			err := v.Validate(fmt.Sprintf(plugin, test.config))
+			if test.wantErr {
+				assert.NotNil(t, err)
+				require.JSONEq(t, test.expectedErr, err.Error())
+			} else {
+				assert.Nil(t, err)
+			}
+		})
+	}
+}
+
+func TestValidator_ValidateMethodsAndHeaders(t *testing.T) {
+	v, err := NewValidator(ValidatorOpts{})
+	assert.Nil(t, err)
+	schema, err := ioutil.ReadFile("testdata/methods-and-headers.lua")
+	assert.Nil(t, err)
+	pluginName, err := v.LoadSchema(string(schema))
+	assert.Nil(t, err)
+	assert.EqualValues(t, "methods-and-headers", pluginName)
+
+	tests := []struct {
+		name        string
+		config      string
+		wantErr     bool
+		expectedErr string
+	}{
+		{
+			name: "valid headers",
+			config: `{
+				"headers": {
+					"my-header": [
+					  "value1"
+					]
+				}
+			}`,
+		},
+		{
+			name: "valid headers",
+			config: `{
+				"headers": {
+					"my-header": [
+					  "value1"
+					],
+					"my-other-header": [
+					  "value2",
+					  "value3"
+					]
+				}
+			}`,
+		},
+		{
+			name: "invalid headers with invalid header name",
+			config: `{
+				"headers": {
+					"my-header!": ["value1"]
+				}
+			}`,
+			wantErr: true,
+			expectedErr: `{"config":{
+				"headers":"bad header name 'my-header!', allowed characters are A-Z, a-z, 0-9, '_', and '-'"
+				}}`,
+		},
```

Comment:
We should test multiples here as well since we have seen issues with validation when things are the first element

Reply:
Apparently Kong doesn't return all the bogus headers at validation time, but only one of them. Also, it doesn't seem to be consistent (meaning, it's not always the last one, or the first one etc...that's probably related to the fact that headers are maps, so it may be due to internal order, returning the first failure it meets)

Last header 
```
$ echo '
name: http-log
config:
    http_endpoint: http://www.konghq.com
    headers:
        my-header:
          - value1
        my-invalid-header!:
          - value1
        my-other-header:
          - value1
        my-other-invalid-header!:
          - value1
' | y2j | http :8001/schemas/plugins/validate

{
    "code": 2,
    "fields": {
        "config": {
            "headers": "bad header name 'my-other-invalid-header!', allowed characters are A-Z, a-z, 0-9, '_', and '-'"
        }
    },
    "message": "schema violation (config.headers: bad header name 'my-other-invalid-header!', allowed characters are A-Z, a-z, 0-9, '_', and '-')",
    "name": "schema violation"
}
```

Not last - not first header

```
$ echo '
name: http-log
config:
    http_endpoint: http://www.konghq.com
    headers:
        my-header:
          - value1
        my-invalid-header!:
          - value1
        my-other-header:
          - value1
        my-other-invalid-header!:
          - value1
        my-yet-another-header:
          - value1
        my-yet-another-invalid-header!:
          - value1
' | y2j | http :8001/schemas/plugins/validate

{
    "code": 2,
    "fields": {
        "config": {
            "headers": "bad header name 'my-other-invalid-header!', allowed characters are A-Z, a-z, 0-9, '_', and '-'"
        }
    },
    "message": "schema violation (config.headers: bad header name 'my-other-invalid-header!', allowed characters are A-Z, a-z, 0-9, '_', and '-')",
    "name": "schema violation"
}
```

Reply:
https://github.com/Kong/goks/pull/14/commits/299ba90d87aa1de282910d3a9a4310ce92e27ff3